frontend/src/components/SystemSubmitDrawer/index.tsx

```diff
@@ -40,6 +41,71 @@ enum State {
   other,
 }
 
+function hasCommonString(one: string, other: string): boolean {
+  return isSubstring(one, other) || isSubstring(other, one);
+}
+function isSubstring(query: string, string: string): boolean {
+  return string.toLowerCase().includes(query.toLocaleLowerCase());
+}
+
+type LangScore = {
+  lang: LanguageCode;
+
+  // The score of the language code, used for ranking during search
+  // Higher value ones are better results and will be displayed earlier in a list
+  value: number;
+};
+
+/**
+ * Given a query, search in a given list of language codes to get
+ * the matching ones. The query will try to match all fields
+ * of a language code with different priority
+ *
+ * @param query a query string used to match all fields
+ * @param data a list of language codes to search from
+ * @returns a filtered list ordered (descending) by score
+ */
+const filterFunc: FilterFunc<LanguageCode> = (query, data) => {
```

Comment:
I think this function can be simplified by using a `Map` so you don't need to keep `added` and `scores` in sync. You can also maybe remove `LangScore` if you have a `Map` of `lang -> score`.